File: java/Samples/Biometrics/VB/SimpleFingersSampleVB/SegmentFingerprints.vb
Imports System
Imports System.Drawing
Imports System.Linq
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Images

Partial Public Class SegmentFingerprints
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()

		OpenFileDialog.Filter = NImages.GetOpenFileFilterString(True, True)
		OpenFileDialog.Filter = OpenFileDialog.Filter

		lbPosition.Items.Add(NFPosition.PlainLeftFourFingers)
		lbPosition.Items.Add(NFPosition.PlainRightFourFingers)
		lbPosition.Items.Add(NFPosition.PlainThumbs)
		lbPosition.Items.Add(NFPosition.LeftLittle)
		lbPosition.Items.Add(NFPosition.LeftRing)
		lbPosition.Items.Add(NFPosition.LeftMiddle)
		lbPosition.Items.Add(NFPosition.LeftIndex)
		lbPosition.Items.Add(NFPosition.LeftThumb)
		lbPosition.Items.Add(NFPosition.RightThumb)
		lbPosition.Items.Add(NFPosition.RightIndex)
		lbPosition.Items.Add(NFPosition.RightMiddle)
		lbPosition.Items.Add(NFPosition.RightRing)
		lbPosition.Items.Add(NFPosition.RightLittle)
		lbPosition.SelectedIndex = 0
	End Sub

#End Region

#Region "Private fields"

	Private _biometricClient As NBiometricClient
	Private _subject As NSubject
	Private _image As NImage

#End Region

#Region "Public properties"

	Public Property BiometricClient() As NBiometricClient
		Get
			Return _biometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			_biometricClient = value
		End Set
	End Property

#End Region

#Region "Private form events"

	Private Sub ClearSegmentInfo()
		pictureBox4.Image = Nothing
		pictureBox3.Image = pictureBox4.Image
		pictureBox2.Image = pictureBox3.Image
		pictureBox1.Image = pictureBox2.Image
		lbPosition4.Text = "Position:"
		lbPosition3.Text = lbPosition4.Text
		lbPosition2.Text = lbPosition3.Text
		lbPosition1.Text = lbPosition2.Text
		lbQuality4.Text = "Quality:"
		lbQuality3.Text = lbQuality4.Text
		lbQuality2.Text = lbQuality3.Text
		lbQuality1.Text = lbQuality2.Text
		lbClass4.Text = "Class:"
		lbClass3.Text = lbClass4.Text
		lbClass2.Text = lbClass3.Text
		lbClass1.Text = lbClass2.Text
	End Sub

	Private Sub Segment()
		If _image Is Nothing Then
			Return
		End If
		_subject = New NSubject()
		Dim finger = New NFinger With {.Image = _image}
		_subject.Fingers.Add(finger)
		originalFingerView.Finger = finger
		finger.Position = CType(lbPosition.SelectedItem, NFPosition)
		_subject.MissingFingers.Clear()
		For Each o As Object In chlbMissing.CheckedItems
			_subject.MissingFingers.Add(CType(o, NFPosition))
		Next o
		_biometricClient.FingersDeterminePatternClass = True
		_biometricClient.FingersCalculateNfiq = True
		Dim task As NBiometricTask = _biometricClient.CreateTask(NBiometricOperations.Segment Or NBiometricOperations.CreateTemplate Or NBiometricOperations.AssessQuality, _subject)
		_biometricClient.BeginPerformTask(task, AddressOf OnSegmentCompleted, Nothing)
	End Sub

	Private Sub OnSegmentCompleted(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnSegmentCompleted), r)
		Else
			Dim task As NBiometricTask = _biometricClient.EndPerformTask(r)
			Dim status As NBiometricStatus = task.Status
			If task.Error IsNot Nothing Then
				Utils.ShowException(task.Error)
			End If
			lblStatus.Text = String.Format("Segmentation status: {0}", status)
			lblStatus.ForeColor = If(status = NBiometricStatus.Ok, Color.Green, Color.Red)
			ShowSegments()
			btnSaveImages.Enabled = status = NBiometricStatus.Ok
		End If
	End Sub

	Private Sub ShowSegments()
		Dim segmentsCount As Integer = _subject.Fingers.Count - 1
		If segmentsCount > 0 AndAlso _subject.Fingers(1).Status = NBiometricStatus.Ok Then
			SetSegmentInfo(_subject.Fingers(1), lbPosition1, lbQuality1, lbClass1, pictureBox1)
		End If
		If segmentsCount > 1 AndAlso _subject.Fingers(2).Status = NBiometricStatus.Ok Then
			SetSegmentInfo(_subject.Fingers(2), lbPosition2, lbQuality2, lbClass2, pictureBox2)
		End If
		If segmentsCount > 2 AndAlso _subject.Fingers(3).Status = NBiometricStatus.Ok Then
			SetSegmentInfo(_subject.Fingers(3), lbPosition3, lbQuality3, lbClass3, pictureBox3)
		End If
		If segmentsCount > 3 AndAlso _subject.Fingers(4).Status = NBiometricStatus.Ok Then
			SetSegmentInfo(_subject.Fingers(4), lbPosition4, lbQuality4, lbClass4, pictureBox4)
		End If
	End Sub

	Private Sub SetSegmentInfo(ByVal finger As NFinger, ByVal position As Label, ByVal quality As Label, ByVal patternClass As Label, ByVal pictureBox As PictureBox)
		Dim attributes As NFAttributes = finger.Objects.FirstOrDefault()
		position.Text = "Position: " & finger.Position.ToString()
		If attributes IsNot Nothing Then
			quality.Text = "Quality: " & attributes.NfiqQuality.ToString()
			patternClass.Text = "Class: " & attributes.PatternClass.ToString()
		End If
		pictureBox.Image = finger.Image.ToBitmap()
	End Sub

	Private Sub PrepareForSegment()
		btnSaveImages.Enabled = False
		lblStatus.Text = String.Empty
		ClearSegmentInfo()
	End Sub

	Private Sub OpenButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles openButton.Click
		PrepareForSegment()
		If openFileDialog.ShowDialog() = DialogResult.OK Then
			Try
				_image = NImage.FromFile(openFileDialog.FileName)
				Segment()
			Catch ex As Exception
				Utils.ShowException(ex)
				lblStatus.Text = "Segmentation status: Error"
			End Try
		End If
	End Sub

	Private Sub LbPositionSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles lbPosition.SelectedIndexChanged
		Dim position = CType(lbPosition.SelectedItem, NFPosition)
		chlbMissing.Items.Clear()
		If (Not NBiometricTypes.IsPositionSingleFinger(position)) Then
			For Each item As NFPosition In NBiometricTypes.GetPositionAvailableParts(position, Nothing)
				chlbMissing.Items.Add(item)
			Next item
		End If
	End Sub

	Private Sub BtnSaveImagesClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSaveImages.Click
		If folderBrowserDialog.ShowDialog() = DialogResult.OK Then
			Try
				For i As Integer = 1 To _subject.Fingers.Count - 1
					If _subject.Fingers(i).Status = NBiometricStatus.Ok Then
						Dim name As String = String.Format("finger{0} {1}{2}", i, _subject.Fingers(i).Position.ToString(), ".png")
						Dim path As String = System.IO.Path.Combine(folderBrowserDialog.SelectedPath, name)
						_subject.Fingers(i).Image.Save(path)
					End If
				Next i
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub SegmentButtonClick(ByVal sender As Object, ByVal e As EventArgs) Handles segmentButton.Click
		If _image IsNot Nothing Then
			PrepareForSegment()
			Segment()
		Else
			MessageBox.Show(String.Format("No image selected!"), Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
		End If
	End Sub

#End Region

End Class

File: java/Samples/Biometrics/CS/SimpleFingersSampleCS/SegmentFingerprints.cs
using System;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Images;

namespace Neurotec.Samples
{
	public partial class SegmentFingerprints : UserControl
	{
		#region Public constructor

		public SegmentFingerprints()
		{
			InitializeComponent();

			openFileDialog.Filter = openFileDialog.Filter = NImages.GetOpenFileFilterString(true, true);

			lbPosition.Items.Add(NFPosition.PlainLeftFourFingers);
			lbPosition.Items.Add(NFPosition.PlainRightFourFingers);
			lbPosition.Items.Add(NFPosition.PlainThumbs);
			lbPosition.Items.Add(NFPosition.LeftLittle);
			lbPosition.Items.Add(NFPosition.LeftRing);
			lbPosition.Items.Add(NFPosition.LeftMiddle);
			lbPosition.Items.Add(NFPosition.LeftIndex);
			lbPosition.Items.Add(NFPosition.LeftThumb);
			lbPosition.Items.Add(NFPosition.RightThumb);
			lbPosition.Items.Add(NFPosition.RightIndex);
			lbPosition.Items.Add(NFPosition.RightMiddle);
			lbPosition.Items.Add(NFPosition.RightRing);
			lbPosition.Items.Add(NFPosition.RightLittle);
			lbPosition.SelectedIndex = 0;
		}

		#endregion

		#region Private fields

		private NBiometricClient _biometricClient;
		private NSubject _subject;
		private NImage _image;

		#endregion

		#region Public properties

		public NBiometricClient BiometricClient
		{
			get { return _biometricClient; }
			set { _biometricClient = value; }
		}

		#endregion

		#region Private form events

		private void ClearSegmentInfo()
		{
			pictureBox1.Image = pictureBox2.Image = pictureBox3.Image = pictureBox4.Image = null;
			lbPosition1.Text = lbPosition2.Text = lbPosition3.Text = lbPosition4.Text = @"Position:";
			lbQuality1.Text = lbQuality2.Text = lbQuality3.Text = lbQuality4.Text = @"Quality:";
			lbClass1.Text = lbClass2.Text = lbClass3.Text = lbClass4.Text = @"Class:";
		}

		private void Segment()
		{
			if (_image == null) return;
			_subject = new NSubject();
			var finger = new NFinger { Image = _image };
			_subject.Fingers.Add(finger);
			originalFingerView.Finger = finger;
			finger.Position = (NFPosition)lbPosition.SelectedItem;
			_subject.MissingFingers.Clear();
			foreach (Object o in chlbMissing.CheckedItems)
			{
				_subject.MissingFingers.Add((NFPosition)o);
			}
			_biometricClient.FingersDeterminePatternClass = true;
			_biometricClient.FingersCalculateNfiq = true;
			NBiometricTask task = _biometricClient.CreateTask(NBiometricOperations.Segment | NBiometricOperations.CreateTemplate | NBiometricOperations.AssessQuality, _subject);
			_biometricClient.BeginPerformTask(task, OnSegmentCompleted, null);
		}

		private void OnSegmentCompleted(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnSegmentCompleted), r);
			}
			else
			{
				NBiometricTask task = _biometricClient.EndPerformTask(r);
				NBiometricStatus status = task.Status;
				if (task.Error != null) Utils.ShowException(task.Error);
				lblStatus.Text = string.Format("Segmentation status: {0}", status);
				lblStatus.ForeColor = status == NBiometricStatus.Ok ? Color.Green : Color.Red;
				ShowSegments();
				btnSaveImages.Enabled = status == NBiometricStatus.Ok;
			}
		}

		private void ShowSegments()
		{
			int segmentsCount = _subject.Fingers.Count - 1;
			if (segmentsCount > 0 && _subject.Fingers[1].Status == NBiometricStatus.Ok)
			{
				SetSegmentInfo(_subject.Fingers[1], lbPosition1, lbQuality1, lbClass1, pictureBox1);
			}
			if (segmentsCount > 1 && _subject.Fingers[2].Status == NBiometricStatus.Ok)
			{
				SetSegmentInfo(_subject.Fingers[2], lbPosition2, lbQuality2, lbClass2, pictureBox2);
			}
			if (segmentsCount > 2 && _subject.Fingers[3].Status == NBiometricStatus.Ok)
			{
				SetSegmentInfo(_subject.Fingers[3], lbPosition3, lbQuality3, lbClass3, pictureBox3);
			}
			if (segmentsCount > 3 && _subject.Fingers[4].Status == NBiometricStatus.Ok)
			{
				SetSegmentInfo(_subject.Fingers[4], lbPosition4, lbQuality4, lbClass4, pictureBox4);
			}
		}

		private void SetSegmentInfo(NFinger finger, Label position, Label quality, Label patternClass, PictureBox pictureBox)
		{
			NFAttributes attributes = finger.Objects.FirstOrDefault();
			position.Text = @"Position: " + finger.Position;
			if (attributes != null)
			{
				quality.Text = @"Quality: " + attributes.NfiqQuality;
				patternClass.Text = @"Class: " + attributes.PatternClass;
			}
			pictureBox.Image = finger.Image.ToBitmap();
		}

		private void PrepareForSegment()
		{
			btnSaveImages.Enabled = false;
			lblStatus.Text = String.Empty;
			ClearSegmentInfo();
		}

		private void OpenButtonClick(object sender, EventArgs e)
		{
			PrepareForSegment();
			if (openFileDialog.ShowDialog() == DialogResult.OK)
			{
				try
				{
					_image = NImage.FromFile(openFileDialog.FileName);
					Segment();
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
					lblStatus.Text = @"Segmentation status: Error";
				}
			}
		}

		private void LbPositionSelectedIndexChanged(object sender, EventArgs e)
		{
			var position = (NFPosition)lbPosition.SelectedItem;
			chlbMissing.Items.Clear();
			if (!NBiometricTypes.IsPositionSingleFinger(position))
			{
				foreach (NFPosition item in NBiometricTypes.GetPositionAvailableParts(position, null))
				{
					chlbMissing.Items.Add(item);
				}
			}
		}

		private void BtnSaveImagesClick(object sender, EventArgs e)
		{
			if (folderBrowserDialog.ShowDialog() == DialogResult.OK)
			{
				try
				{
					for (int i = 1; i < _subject.Fingers.Count; i++)
					{
						if (_subject.Fingers[i].Status == NBiometricStatus.Ok)
						{
							string name = string.Format("finger{0} {1}{2}", i, _subject.Fingers[i].Position.ToString(), ".png");
							string path = System.IO.Path.Combine(folderBrowserDialog.SelectedPath, name);
							_subject.Fingers[i].Image.Save(path);
						}
					}
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void SegmentButtonClick(object sender, EventArgs e)
		{
			if (_image != null)
			{
				PrepareForSegment();
				Segment();
			}
			else MessageBox.Show(string.Format("No image selected!"), Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
		}

		#endregion
	}
}
